### ISSUE TITLE
Cannot assign dict to TypedDict subclass

### ISSUE THREAD

**Issue by apollo13** [microsoft/pyright#4533]
**Describe the bug**
Running `pyright` on my Codebase yields the following issue:
```
test.py:17:32 - error: Expression of type "dict[str, str | bytes | bool]" cannot be assigned to declared type "HTTPResponseBodyEvent" (reportGeneralTypeIssues)
```

**To Reproduce**
Install `pyright==1.1.291`, `hypercorn==0.14.3` and run pyright on this file:
```
from typing import Literal, TypedDict

from hypercorn.typing import HTTPResponseBodyEvent


class Test(TypedDict):
    type: Literal["http.response.body"]
    body: bytes
    more_body: bool


test: Test = {
    "type": "http.response.body",
    "body": b"asd",
    "more_body": False,
}
test2: HTTPResponseBodyEvent = {
    "type": "http.response.body",
    "body": b"asd",
    "more_body": False,
}
```

**Expected behavior**
I would have expected that `test` and `test2` both validate fine. Please note that `Test` is literally the same definition as `HTTPResponseBodyEvent`

**VS Code extension or command-line**
Pyright commandline installed via pip

**Comment by erictraut** [comment]
This is due tot he way the `hypercorn` library is importing the `TypedDict` symbol. It's effectively importing it from two different modules: `typing` and `typing_extensions`, and it's doing so using a dynamic construct (a try/except block).

```python
try:
    from typing import Literal, Protocol, TypedDict
except ImportError:
    from typing_extensions import Literal, Protocol, TypedDict  # type: ignore
```

This is not friendly for static analyzers like pyright. The recommended approach is:
```python
import sys
if sys.version_info >= (3, 8):
    from typing import TypedDict
else:
    from typing_extensions import TypedDict
```

However, I've seen this pattern show up in several libraries now, so I've added some additional code to pyright that incorporates some heuristics to handle this. I generally try to avoid heuristics because they tend to add complexity and fail in edge cases, but I think it's justified in this case.

With the change in place, this library works as expected. This will be included in the next release of pyright.

**Comment by apollo13** [comment]
Hi Eric, thank you for your quick response and the fix. I agree that this situation might be tricky for static analyzers, but the `try/except` feels more pythonic. Thanks very much!

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.292, which I just published. It will also be included in a future release of pylance.
